[Inductor-CPU] int8 WoQ concat linear

sanchitintel opened the pull request:
### Summary

int8 WoQ GEMM concat linear optimization pertaining to the same activation applied to 3 sets of weights of the same shape.


### Perf data 

GPT-J 128 input tokens, 128 output tokens.
32 physical cores of one socket of Intel(R) Xeon(R) 6972P (Xeon Gen 5). tcmalloc & Intel OpenMP were preloaded.

| May 8 nightly first token latency | First token latency with this implementation | Rest token latency with May 8 nightly | Rest token latency with this implementation combined with #149373  |
|---|---|---|---|
|202 ms | 190 ms | 33 ms | 30 ms|




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @chunyuan-w @leslie-fang-intel @Xia-Weiwen

sanchitintel committed:
[skip ci] WoQ concat linear Change 1

sanchitintel committed:
[skip ci] Change 2/3

sanchitintel committed:
[Change 3/3] Add concat linear

sanchitintel committed:
Fix all corner-cases

sanchitintel committed:
Reduce duplicate code

sanchitintel committed:
Remove redundant code

sanchitintel committed:
Merge branch 'pytorch:main' into woq_concat_linear

sanchitintel committed:
Merge branch 'pytorch:main' into woq_concat_linear

sanchitintel committed:
Merge branch 'pytorch:main' into woq_concat_linear

sanchitintel committed:
Merge branch 'pytorch:main' into woq_concat_linear

sanchitintel committed:
Better variables names in UT

sanchitintel committed:
Disable replacement on all devices except CPU

sanchitintel committed:
Merge branch 'pytorch:main' into woq_concat_linear

sanchitintel committed:
Add torchao implementation based UT

sanchitintel committed:
Fix lint

Make suggested changes

Co-authored-by: Anthony Shoumikhin <anthony@shoumikh.in>

sanchitintel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -2185,7 +2324,8 @@ def _register_qlinear_weight_prepack():
     # refer to:
-    # https://github.com/pytorch/pytorch/blob/80c07df659362a95da7cd4f3ec367abfdace38c4/torch/_decomp/decompositions.py#L3965-L3968
+    # https://github.com/pytorch/pytorch/blob/
+    # 80c07df659362a95da7cd4f3ec367abfdace38c4/torch/_decomp/decompositions.py#L3965-L3968
```
Thank you! :)

sanchitintel commented on torch/_inductor/fx_passes/freezing_patterns.py:
```diff
@@ -121,10 +121,55 @@ def register_binary_folding_pattern(pattern, extra_check=_return_true):
+        if not is_cpu:
+            # Currently, this pattern is only supported on CPU
+            return False
+        elif is_cpu and not config.cpp.enable_concat_linear:
```
Sorry, that looks better

sanchitintel commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -1094,6 +1143,73 @@ def fn(match):
+        # We call stable_topological_sort here as a workaround.
+        stable_topological_sort(match.graph)
+        with match.graph.inserting_before(user_of_scaling_node):
+            new_cat_node = match.graph.call_function(
```
Yes

leslie-fang-intel replied:
Thanks. Then this PR should look good to me.

sanchitintel committed:
Refactor

Was half-asleep when I made a silly change :(

sanchitintel committed:
Fix lint

sanchitintel committed:
Fix typo for fixing lint

sanchitintel commented:
Hi @jansel, can you please help review this PR as per your convenience & availability? 
It requires approval from a core reviewer for landing, so I added you as a reviewer.

Thank you!

sanchitintel commented:
Hi @ezyang, please advise if you'd be able to help land this PR. Thanks!

@leslie-fang-intel has landing permissions for all the files being modified in this PR, except for `torch/_inductor/fx_passes/freezing_patterns.py`.

sanchitintel committed:
Restrict pattern replacement to Inductor config cpp.enable_concat_linear being true

`config.cpp.enable_concat_linear` was not being checked in `torch/_inductor/fx_passes/quantization.py`.

Note that we didn't make such a change in `_mkldnn_weight_pack_init` because of a corner-case - although unlikely, it's possible that some workload's process may enable `config.cpp.enable_concat_linear` after `_mkldnn_weight_pack_init` has already been called once. Such a scenario could happen if a single process would compile forwards of multiple models, and would want to enable `config.cpp.enable_concat_linear` for some models, but not others.

sanchitintel commented:
@pytorchbot merge

sanchitintel commented:
The failed CI job [also failed on many other PRs today](https://hud.pytorch.org/failure?name=inductor%20%2F%20linux-jammy-cpu-py3.9-gcc11-inductor%20%2F%20test%20(inductor_torchbench_cpu_smoketest_perf%2C%201%2C%201%2C%20linux.24xl.spr-metal%2C%20unstable)&jobName=linux-jammy-cpu-py3.9-gcc11-inductor%20%2F%20test%20(inductor_torchbench_cpu_smoketest_perf%2C%201%2C%201%2C%20linux.24xl.spr-metal%2C%20unstable)&failureCaptures=Process%20completed%20with%20exit%20code%201.) with the same root-cause, and is thus unrelated to this PR.
